x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -153,26 +155,31 @@ public void authorizeClusterAction(RequestInfo requestInfo, AuthorizationInfo au
     boolean checkSameUserPermissions(String action, TransportRequest request, Authentication authentication) {
         final boolean actionAllowed = SAME_USER_PRIVILEGE.test(action);
         if (actionAllowed) {
-            if (request instanceof UserRequest == false) {
-                assert false : "right now only a user request should be allowed";
-                return false;
-            }
-            UserRequest userRequest = (UserRequest) request;
-            String[] usernames = userRequest.usernames();
-            if (usernames == null || usernames.length != 1 || usernames[0] == null) {
-                assert false : "this role should only be used for actions to apply to a single user";
+            if (request instanceof UserRequest) {
+                UserRequest userRequest = (UserRequest) request;
+                String[] usernames = userRequest.usernames();
+                if (usernames == null || usernames.length != 1 || usernames[0] == null) {
+                    assert false : "this role should only be used for actions to apply to a single user";
+                    return false;
+                }
+                final String username = usernames[0];
+                final boolean sameUsername = authentication.getUser().principal().equals(username);
+                if (sameUsername && ChangePasswordAction.NAME.equals(action)) {
+                    return checkChangePasswordAction(authentication);
+                }
+
+                assert AuthenticateAction.NAME.equals(action) || HasPrivilegesAction.NAME.equals(action)
+                        || GetUserPrivilegesAction.NAME.equals(action)
+                        || sameUsername == false : "Action '" + action + "' should not be possible when sameUsername=" + sameUsername;
+                return sameUsername;
+            } else if (request instanceof GetMyApiKeyRequest) {
+                if (authentication.getAuthenticatedBy().getType().equals("_es_api_key")) {
+                    return true;
+                }
+            } else {
+                assert false : "only a user request or get my API key request should be allowed";
```

Comment:
I see, `GetMyApiKey` is authorized implicitly, but create and invalidate are not.
But, isn't the condition `if (authentication.getAuthenticatedBy().getType().equals("_es_api_key"))` checking that the authn has been performed using the API key?
If this is the case, what would be the benefit of the client that authn with the API key to get his own api keys, but if he authn with the password to be rejected? Assuming she will be rejected, because it is unlikely, since to "get" them they must have been created by the "self" and the privilege to create grants get as well.